### PR TITLE
dev_localSearch_lhcSampling_nb101dictCreation

### PR DESCRIPTION
**Pull request by Stargazer1998**
Add a script for the creation of Nas101 dict. creator and helper functions to be used

## Describe your changes
Common functions to be used for LHC Sampler and nb101 dict creation
nb101 dictionary creator script 
LCH Sampler Script 
modify local search to work with sampled points and nb101 dict. 
(local search tests are not active for now )
## What is the related task number?
-
## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: src/tum_dlr_automl_for_eo/scripts/nb101_dict_creator.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#       path to save the nb101_dict file
+
+path_to_read_data = '.../nasbench_full.tfrecord'
+path_to_save_dictionary = ".../tum-dlr-automl-for-eo/data/nb101_dict"
+
+data_format = 'tfrecord'  # data format to read
+
+
+def create_encoded_arc_file():
+    nasbench = api.NASBench(path_to_read_data, data_format = data_format)
```
Comment (emreds) [inline src/tum_dlr_automl_for_eo/scripts/nb101_dict_creator.py]: Why we are iterating over all the architectures in nasbench here?

File: src/tum_dlr_automl_for_eo/utils/helper_functions.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        elif op == MAXPOOL3X3:
+            new_ops = new_ops + [op + "_" + str(mp3x3)]
+            mp3x3 = mp3x3 + 1
+        else:
+            new_ops = new_ops + [op]
+    return new_ops
+
+
+def sample_single_op(NUM_VERTICES, ALLOWED_OPS):
+    x = lhs(NUM_VERTICES, samples=1)[0]
```
Comment (emreds) [inline src/tum_dlr_automl_for_eo/utils/helper_functions.py]: I could not find the definition of the `lhc` function. Could you show it to me? @Stargazer1998

File: src/tum_dlr_automl_for_eo/scripts/nb101_dict_creator.py (hunk 1 above)
Comment (Stargazer1998) [reply src/tum_dlr_automl_for_eo/scripts/nb101_dict_creator.py]: We would like to create a kind of look-up table for each nb101 architecture, as {encoded_arc: nb101_arc}. Thus, we do not need to decode again during neighbour search.

File: src/tum_dlr_automl_for_eo/utils/helper_functions.py (hunk 2 above)
Comment (Stargazer1998) [reply src/tum_dlr_automl_for_eo/utils/helper_functions.py]: Upps, I should have been more clear. It is a function from pyDOE lib.

Comment (emreds) [reply src/tum_dlr_automl_for_eo/utils/helper_functions.py]: Hmm then we can import it as `import pyDOE` and use it like `pyDOE.lhs()`. 
Could you add the new required packages to the requirements.txt also?

File: src/tum_dlr_automl_for_eo/scripts/nb101_dict_creator.py (hunk 1 above)
Comment (emreds) [reply src/tum_dlr_automl_for_eo/scripts/nb101_dict_creator.py]: I see makes sense :+1:

File: src/tum_dlr_automl_for_eo/utils/helper_functions.py (hunk 2 above)
Comment (Stargazer1998) [reply src/tum_dlr_automl_for_eo/utils/helper_functions.py]: sure

**Review by emreds — APPROVED** [review]

File: src/tum_dlr_automl_for_eo/utils/helper_functions.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    return current_mat, current_op, current_spec
+
+
+def encode_matrix(adj_matrix, ops):
+    enc_matrix = np.zeros((len(CODING), len(CODING)))
+    pos = [CODING.index(op) for op in ops]
+    trans = dict()
+    for i, ix in enumerate(pos):
+        trans[i] = ix
+    i, j = np.nonzero(adj_matrix)
```
Comment (emreds) [inline src/tum_dlr_automl_for_eo/utils/helper_functions.py:120]: Hmm maybe we should also `import numpy as np` flake8 gives an error about the not imported package

**Comment by JahodaPaul** [comment]
So in general I think it's good, my only question is about using LHS sampling. In Latin hypercube sampling one must first decide how many sample points to use and for each sample point remember in which row and column the sample point was taken. However, in the current implementation, we sample the LHS function one by one, which I think maybe defeats the purpose of using LHS sampling.

**Review by JahodaPaul — COMMENTED** [review]
we shouldn't sample lhs 1 by 1 as it defeats the purpose of LHS sampling algorithm

**Comment by Stargazer1998** [comment]
> we shouldn't sample lhs 1 by 1 as it defeats the purpose of LHS sampling algorithm

  Thanks for comments @JahodaPaul. Actually, this function has been provided by René, it would be better to discuss with him.

**Comment by emreds** [comment]
If you have any solutions please implement it now because it's kind of urgent @JahodaPaul  :smile:

**Comment by JahodaPaul** [comment]
> If you have any solutions please implement it now because it's kind of urgent @JahodaPaul smile

I think after a discussion with Safa, he now knows how to implement it and if I am not wrong he's working on fixing it.

**Pull request merged**